core/amber/src/test/scala/edu/uci/ics/amber/engine/architecture/scheduling/DefaultCostEstimatorSpec.scala

```diff
@@ -0,0 +1,206 @@
+    val workflowExecution = new WorkflowExecutions
+    workflowExecution.setEid(UInteger.valueOf(1))
+    workflowExecution.setVid(UInteger.valueOf(1))
+    workflowExecution.setUid(UInteger.valueOf(1))
+    workflowExecution.setStatus(3.toByte)
+    workflowExecution.setEnvironmentVersion("test engine")
+    workflowExecution
+  }
+
+  private val headerlessCsvOpStatisticsEntry: WorkflowRuntimeStatistics = {
+    val workflowRuntimeStatistics = new WorkflowRuntimeStatistics
+    workflowRuntimeStatistics.setOperatorId(headerlessCsvOpDesc.operatorIdentifier.id)
+    workflowRuntimeStatistics.setWorkflowId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setExecutionId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setDataProcessingTime(ULong.valueOf(100))
+    workflowRuntimeStatistics.setControlProcessingTime(ULong.valueOf(100))
+    workflowRuntimeStatistics
+  }
+
+  private val keywordOpDescStatisticsEntry: WorkflowRuntimeStatistics = {
+    val workflowRuntimeStatistics = new WorkflowRuntimeStatistics
+    workflowRuntimeStatistics.setOperatorId(keywordOpDesc.operatorIdentifier.id)
+    workflowRuntimeStatistics.setWorkflowId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setExecutionId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setDataProcessingTime(ULong.valueOf(300))
+    workflowRuntimeStatistics.setControlProcessingTime(ULong.valueOf(300))
+    workflowRuntimeStatistics
+  }
+
+  private val sinkOpDescStatisticsEntry: WorkflowRuntimeStatistics = {
+    val workflowRuntimeStatistics = new WorkflowRuntimeStatistics
+    workflowRuntimeStatistics.setOperatorId(sink.operatorIdentifier.id)
+    workflowRuntimeStatistics.setWorkflowId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setExecutionId(UInteger.valueOf(1))
+    workflowRuntimeStatistics.setDataProcessingTime(ULong.valueOf(200))
+    workflowRuntimeStatistics.setControlProcessingTime(ULong.valueOf(200))
+    workflowRuntimeStatistics
+  }
+
+  override protected def beforeAll(): Unit = {
+    initializeDBAndReplaceDSLContext()
+  }
+
+  "DefaultCostEstimator" should "use fallback method when no past statistics are available" in {
+    val workflow = buildWorkflow(
+      List(headerlessCsvOpDesc, keywordOpDesc, sink),
+      List(
+        LogicalLink(
+          headerlessCsvOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0)
+        ),
+        LogicalLink(
+          keywordOpDesc.operatorIdentifier,
+          PortIdentity(0),
+          sink.operatorIdentifier,
+          PortIdentity(0)
+        )
+      ),
+      new WorkflowContext()
+    )
+
+    val costEstimator = new DefaultCostEstimator(
+      workflow.context,
+      CONTROLLER
+    )
+
+    val region = Region(
+      id = RegionIdentity(0),
+      physicalOps = workflow.physicalPlan.operators,
+      physicalLinks = workflow.physicalPlan.links
+    )
```

Comment:
can we add a test case where the region only contains a sub-dag of the workflow? just to make sure it can extract region costs correctly.

Reply:
Added one more test case.